src/Functions/EntropyLeardnedHash.cpp

```diff
@@ -0,0 +1,104 @@
+#include "EntropyLeardnedHash.h"
+#include <algorithm>
+#include <cstring>
+#include <map>
+#include <set>
+#include <unordered_map>
+#include <unordered_set>
+
+namespace
+{
+bool AreAllPartialKeysUnique(const std::vector<EntropyLearnedHashing::Key> & data, const std::vector<size_t> & positions)
+{
+    static std::unordered_set<EntropyLearnedHashing::Key> partial_keys;
+    partial_keys.reserve(data.size());
+    for (const auto & key : data)
+    {
+        if (!partial_keys.insert(EntropyLearnedHashing::getPartialKey(key, positions)).second)
+            return false;
+    }
+    return true;
+}
+
+// NextByte returns position of byte which adds the most entropy and the new entropy
+std::pair<size_t, size_t> NextByte(const std::vector<EntropyLearnedHashing::Key> & keys, size_t max_len, std::vector<size_t> & chosen_bytes)
+{
+    size_t min_collisions = std::numeric_limits<size_t>::max();
+    size_t best_position = 0;
+
+    for (size_t i = 0; i < max_len; ++i)
+    {
+        static std::unordered_map<EntropyLearnedHashing::Key, size_t> count_table;
```

Comment:
Let's not make `count_table` static. This will only lead to issues when NextByte runs in parallel. You could instead pull it out of the loop (l. 29). 

Same for `partial_keys` (l. 13).